### ISSUE TITLE
Setting up a Noray Instance

### ISSUE THREAD

**Issue by Learus** [foxssake/netfox#218]
Hi there.

I've been developing my multiplayer game for several months now using the tomfol.io instance provided and it's been working great. The game is ready to alpha launch feature-wise, and I wanted to move over from your instance to a custom one that I handle myself.

Although I have followed the setup instructions on the noray repository, I can't have it working properly. I managed to connect to the Noray instance and "register_host". I'm getting both an OID and PID successfully. When it comes to the register_remote call, on the other hand, it times out (error 24). Seems like no packets come from the UDP relay.
I wonder if I'm doing something wrong/forgetting something, or something else entirely.

Could you provide some insight on the issue? Perhaps share your setup of the tomfol.io instance so we can replicate it?

As extra info, I've tried hosting the Noray instance using docker and also locally with pnpm. I've also changed the host to IPv4 (0.0.0.0).

Any help would be appreciated!

**Comment by Learus** [comment]
Hey, thanks for the quick reply!

I've tried both hosting it with docker (both locally and in a server), and locally with pnpm start.
I had already assumed I had to expose port 8809 so I had already tried that. I hadn't tried exposing the relay ports as well though.

Sadly, none of the solutions that you propose are working 😞. "network_mode: host" does not seem to be working at all weirdly enough, as in the service doesn't even respond to the "register-host" command anymore. 
As an extra note, exposing 130 ports is indeed slow when docker starts. I wonder if it hinders general performance as well.

Any ideas?

--- 

Regarding the showcasing of the game, sure! Just need to make sure that it's stable and working before I publish the alpha. When that's done, I'll be glad to send a message and move it forward.

**Comment by elementbound** [comment]
Gotcha, in that case let's debug with running through pnpm on localhost - I'm wondering where does the connectivity break. Try running noray with `NORAY_LOGLEVEL=trace` and then connect to it. If you get a `Received UDP relay request`, then the UDP packets arrive to noray, but it fails to relay for whatever reason.

If it doesn't arrive, then it's either a config issue ( noray is listening on a different address/port where the game is sending the packets ), or some mysterious network issue.

Also, do you get the same issue with Forest Brawl? Alternatively, how's your game using the noray addon?

**Comment by Learus** [comment]
Ok, good news!
Turns out that my local docker had issues with "network_mode: host", since I'm using Docker Desktop with WSL. If I try it on the external server, on the other hand, it works properly. So, I guess it was a config issue. I'll keep troubleshooting on my local computer just to figure it out for everybody else having the same issue as me.

Another possible bug I noticed now that I got it working:
1. Run the game via the Godot Editor
2. Connect to noray (I do it automatically for now)
3. Shut the game down using the Stop button in the editor
4. Noray server crashes with ECONNRESET.

If I shut down the game normally (by get_tree().quit(), or by pressing the X button on the main window) everything is fine.
For now, it's not a big issue for dev, because I could just "restart: always" the dev noray server. Nevertheless, it could cause problems in the future. LMK if I should post this as a different issue.

**Comment by elementbound** [comment]
Thanks! Could you please file an issue under noray for the crash?

**Comment by MattStory** [comment]
Hi, I'm using a similar setup (WSL + Docker) and running into the same issue when trying to connect via localhost. Were you able to figure out any solution besides moving to an external server? Thanks!

**Comment by elementbound** [comment]
Hi @MattStory, 

So to confirm, you're running noray locally from Docker, you're able connect to it from your game, get an OID/PID pair, but the game fails at registering the remote address?

Some ideas for troubleshooting:

1. Is traffic getting through to the remote registrar? This is by default on port 8809
    1. If that's working, you should see *"Received UDP relay request"* debug log messages
    2. *Note* that you might need to run noray with `NORAY_LOG_LEVEL=debug` to see these logs
    3. If traffic doesn't get through, check if the remote registrar port is exposed from the container
2. Are the remote address registrations failing?
    1. Check the value of the `noray_remote_registrar_fail` metric at <http://localhost:8891/metrics>
    2. If something somehow fails even more catastrophically, you should see some error messages in the logs
3. Is the relay failing?
    1. If relays get lost for whatever reason, noray will drop their traffic
    2. Check the value of the `noray_relay_drop_count` metric at <http://localhost:8891/metrics>

**Comment by Mondanzo** [comment]
Hope I'm not adding to this unnecessarily, but since I've also struggled to set up a docker instance I noticed some issues in the documentation of the docker instance:

The udp port is never mentioned in the docs and not exposed as an udp port with 8809/udp. Additionally, the docker command shown and the ports set up in the .env.example are vastly different. The udp port range is never exposed at all and shouldn't work as far as I'm aware unless also specifically exposed (which can make creating the instance take a lot longer, like ~142 seconds all of a sudden). I would also suggest setting the TCP connection port and UDP registar port to the same port to save up and avoid any confusion as there is no reason not to use the same port with different protocols.
